From ca4dc1fd07c2d64834ede83d387d4feb2791c185 Mon Sep 17 00:00:00 2001
From: Ping Gao <ping.a.gao@intel.com>
Date: Tue, 20 Dec 2016 09:45:59 +0800
Subject: [PATCH 444/446] vgt: qos: keep fairness round-robin after longest
 unsched guest detected

Longest unsched guest has the highest priority to execute to avoid
TDR, however this mechanism would make some guest lose some chance
to get scheduled. This patch is to keep fairness round-robin to fix
this problem.

Signed-off-by: Ping Gao <ping.a.gao@intel.com>
Reviewed-by: Zheng, Xiao <xiao.zheng@intel.com>
---
 drivers/gpu/drm/i915/vgt/sched.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/vgt/sched.c b/drivers/gpu/drm/i915/vgt/sched.c
index 5df23a5..f1194d3 100644
--- a/drivers/gpu/drm/i915/vgt/sched.c
+++ b/drivers/gpu/drm/i915/vgt/sched.c
@@ -201,8 +201,14 @@ static struct vgt_device *tbs_next_vgt(
 		return pdev->next_sched_vgt;
 
 	next_vgt = vgt_longest_unsched(head);
-	if (next_vgt)
+	if (next_vgt) {
+		/* Move the longest unsched guest after current guest to
+		 * keep fairness round-robin.
+		 */
+		list_del(&next_vgt->list);
+		list_add(&next_vgt->list, &cur_vgt->list);
 		return next_vgt;
+	}
 
 	return vgt_pickup_next(head, cur_vgt);
 }
-- 
1.7.10.4

